packages/config-eslint/moss-lint-plugin/rules/only-valid-token-names.js

```diff
@@ -0,0 +1,109 @@
+// FIXME: temporary solution with test.json file
+import themeInterface from "../test.json" with { type: "json" };
+
+const colorKeysToCssValues = (colors) => {
+  const keys = Object.keys(colors);
+  return keys.map((color) => "--color-" + color.replaceAll(".", "-"));
+};
+
+const VALID_TOKENS = colorKeysToCssValues(themeInterface.colors);
+const SELECTOR_WITH_ARBITRARY_VALUE = /(?:group-)?\w+-\[(?:var\((--[\w-]+)\)|(--[\w-]+))\]/g;
+const ARBITRARY_VALUE = /\[(?:var\((--[\w-]+)\)|(--[\w-]+))\]/;
+
+const getInvalidTokenName = (str) => {
+  const selectors = str.match(SELECTOR_WITH_ARBITRARY_VALUE);
+  if (selectors === null || selectors.length === 0) return false;
+
+  const tokens = selectors.map((className) => {
+    const selector = className.match(ARBITRARY_VALUE);
+    const token = selector[1] || selector[2];
+    return token;
+  });
+
+  return tokens.find((token) => !VALID_TOKENS.includes(token));
+};
+
+const hasInvalidTokens = (str) => {
+  const selectors = str.match(SELECTOR_WITH_ARBITRARY_VALUE);
+
+  if (selectors === null || selectors.length === 0) return false;
+
+  return selectors.some((className) => {
+    const selector = className.match(ARBITRARY_VALUE);
+    const token = selector[1] || selector[2];
+
+    return !VALID_TOKENS.includes(token);
+  });
+};
+
+const getInvalidSelectorLoc = (str, loc) => {
+  const selectors = str.match(SELECTOR_WITH_ARBITRARY_VALUE);
+  if (selectors === null || selectors.length === 0) return false;
+
+  const invalidSelector = selectors.find((className) => {
+    const selector = className.match(ARBITRARY_VALUE);
+    const token = selector[1] || selector[2];
+
```

Comment:
The functions `getInvalidTokenName`, `hasInvalidTokens`, and `getInvalidSelectorLoc` have overlapping logic...

```typescript
const getInvalidTokenName = (str) => {
  let match;
  while ((match = SELECTOR_WITH_ARBITRARY_VALUE.exec(str)) !== null) {
    const { token } = match.groups;
    if (!VALID_TOKENS.has(token)) {
      return token;
    }
  }
  return false;
};

const hasInvalidTokens = (str) => {
  let match;
  while ((match = SELECTOR_WITH_ARBITRARY_VALUE.exec(str)) !== null) {
    const { token } = match.groups;
    if (!VALID_TOKENS.has(token)) {
      return true;
    }
  }
  return false;
};

const getInvalidSelectorLoc = (str, loc) => {
  let match;
  while ((match = SELECTOR_WITH_ARBITRARY_VALUE.exec(str)) !== null) {
    const { token } = match.groups;
    if (!VALID_TOKENS.has(token)) {
      const startIndex = match.index;
      const endIndex = SELECTOR_WITH_ARBITRARY_VALUE.lastIndex;
      const startColumn = loc.start.column + startIndex + 1;
      const endColumn = loc.start.column + endIndex;
      return {
        start: {
          line: loc.start.line,
          column: startColumn,
        },
        end: {
          line: loc.start.line,
          column: endColumn,
        },
      };
    }
  }
  return loc;
};
```

Will this work? (I might have written some code with errors in certain places)

Reply:
This particular code won't. But i agree that the code could be refactored. So I did. 
Now algorithm runs regex on string once, simultaneously extracting all encountering tokens. Than it compares tokens to valid tokens and forms an array of all invalid tokens and their locations. And then reports it to eslint.

The side effect of it's improvment is also that eslint now highlights all the identical errors in the given string(previously it was able to only highlight only the first error in the string)